File: algorithms/huffman_coding/huffman.cpp
/**
 * Huffman Encoding and Decoding Algorithm in C++.
 */

#include <iostream>
#include <string>
#include <queue>
#include <unordered_map>
#include <vector>
#include <cassert>

struct HuffmanNode {
    char ch;
    int freq;
    HuffmanNode* left;
    HuffmanNode* right;

    HuffmanNode(char c, int f, HuffmanNode* l = nullptr, HuffmanNode* r = nullptr)
        : ch(c), freq(f), left(l), right(r) {}
};

struct Compare {
    bool operator()(HuffmanNode* l, HuffmanNode* r) {
        return l->freq > r->freq;
    }
};

void generate_codes(HuffmanNode* root, const std::string& str, std::unordered_map<char, std::string>& huffmanCode) {
    if (!root) return;
    if (!root->left && !root->right) {
        huffmanCode[root->ch] = str;
    }
    generate_codes(root->left, str + "0", huffmanCode);
    generate_codes(root->right, str + "1", huffmanCode);
}

HuffmanNode* build_tree(const std::string& text) {
    std::unordered_map<char, int> freq;
    for (char ch : text) freq[ch]++;

    std::priority_queue<HuffmanNode*, std::vector<HuffmanNode*>, Compare> pq;
    for (auto pair : freq) {
        pq.push(new HuffmanNode(pair.first, pair.second));
    }

    while (pq.size() > 1) {
        HuffmanNode* left = pq.top(); pq.pop();
        HuffmanNode* right = pq.top(); pq.pop();
        int sum = left->freq + right->freq;
        pq.push(new HuffmanNode('\0', sum, left, right));
    }

    return pq.top();
}

std::string decode(HuffmanNode* root, const std::string& encoded) {
    std::string ans = "";
    HuffmanNode* curr = root;
    for (char bit : encoded) {
        if (bit == '0') curr = curr->left;
        else curr = curr->right;

        if (!curr->left && !curr->right) {
            ans += curr->ch;
            curr = root;
        }
    }
    return ans;
}

int main() {
    std::string text = "this is an example for a huffman encoding demonstration";
    HuffmanNode* root = build_tree(text);

    std::unordered_map<char, std::string> huffmanCode;
    generate_codes(root, "", huffmanCode);

    std::string encoded = "";
    for (char ch : text) encoded += huffmanCode[ch];

    std::string decoded = decode(root, encoded);
    assert(decoded == text);

    std::cout << "[C++ Huffman] Message decoded successfully: " << decoded.substr(0, 20) << "..." << std::endl;
    return 0;
}

File: algorithms/huffman_coding/Huffman.java
import java.util.*;

public class Huffman {

    private static class Node implements Comparable<Node> {
        final char ch;
        final int freq;
        final Node left, right;

        Node(char ch, int freq, Node left, Node right) {
            this.ch = ch;
            this.freq = freq;
            this.left = left;
            this.right = right;
        }

        boolean isLeaf() {
            return left == null && right == null;
        }

        @Override
        public int compareTo(Node o) {
            return Integer.compare(this.freq, o.freq);
        }
    }

    private static void buildCode(Map<Character, String> map, Node x, String s) {
        if (!x.isLeaf()) {
            buildCode(map, x.left, s + '0');
            buildCode(map, x.right, s + '1');
        } else {
            map.put(x.ch, s);
        }
    }

    public static Node buildTree(String text) {
        Map<Character, Integer> freq = new HashMap<>();
        for (char c : text.toCharArray()) freq.put(c, freq.getOrDefault(c, 0) + 1);

        PriorityQueue<Node> pq = new PriorityQueue<>();
        for (Map.Entry<Character, Integer> e : freq.entrySet()) {
            pq.add(new Node(e.getKey(), e.getValue(), null, null));
        }

        while (pq.size() > 1) {
            Node left = pq.poll();
            Node right = pq.poll();
            pq.add(new Node('\0', left.freq + right.freq, left, right));
        }

        return pq.poll();
    }

    public static String decode(Node root, String encoded) {
        StringBuilder sb = new StringBuilder();
        Node curr = root;
        for (char bit : encoded.toCharArray()) {
            curr = (bit == '0') ? curr.left : curr.right;
            if (curr.isLeaf()) {
                sb.append(curr.ch);
                curr = root;
            }
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        String msg = "this is an example for a huffman encoding demonstration";
        Node root = buildTree(msg);

        Map<Character, String> codeMap = new HashMap<>();
        buildCode(codeMap, root, "");

        StringBuilder encoded = new StringBuilder();
        for (char c : msg.toCharArray()) encoded.append(codeMap.get(c));

        String decoded = decode(root, encoded.toString());
        if (!decoded.equals(msg)) {
            throw new AssertionError("Huffman decode mismatch");
        }

        System.out.println("[Java Huffman] Successfully encoded and decoded: " + decoded.length() + " chars.");
    }
}
